### PR TITLE
config related metrics

### PR DESCRIPTION
**Pull request by wasaga**
## Summary

adds two new metrics: 

- `pomerium_config_db_errors{config_id="dashboard"}` - amount of errors, including routes skipped while parsing this config 
- `pomerium_config_db_version{config_id="dashboard"}` - current databroker version that was applied

## Related issues

https://github.com/pomerium/pomerium-console/issues/708

## Checklist

- [x] reference any related issues
- [ ] updated docs
- [ ] updated unit tests
- [ ] updated UPGRADING.md
- [ ] add appropriate tag (`improvement` / `bug` / etc)
- [x] ready for review

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -10,7 +10,10 @@
       "args": [
         "-config",
         "${workspaceRoot}/.config.yaml"
-      ]
+      ],
+      "env": {
+        "PATH": "${workspaceRoot}/bin/.getenvoy/builds/standard/1.17.1/darwin/bin/:${env:PATH}"
+      }
```
Comment (calebdoxsey) [inline .vscode/launch.json:16]: I don't think we want this?

File: pkg/grpc/databroker/syncer.go [hunk 2]
```diff
@@ -148,6 +155,12 @@ func (syncer *Syncer) sync(ctx context.Context) error {
 			return err
 		}
 
+		syncer.log().Info().
```
Comment (calebdoxsey) [inline pkg/grpc/databroker/syncer.go:158]: I think this should be `Debug`?

File: pkg/metrics/constants.go [hunk 3]
```diff
@@ -19,6 +19,14 @@ const (
...
 	// ConfigChecksumDecimal should only be used to compare config on a single node, it will be different in multi-node environment
 	ConfigChecksumDecimal = "config_checksum_decimal"
+	// ConfigDbVersion sets currently loaded databroker config version config_db_version{service="service",id="config_id"}
+	ConfigDbVersion = "config_db_version"
+	// ConfigDbVersionHelp
+	ConfigDbVersionHelp = "databroker current config record version"
+	// ConfigDbErrors sets number of errors while parsing current config that were tolerated
+	ConfigDbErrors = "config_db_errors"
+	// ConfigDbErrorsHelp
+	ConfigDbErrorsHelp = "amount of errors observed while applying databroker config; -1 if validation failed and was rejected altogether"
```
Comment (calebdoxsey) [inline pkg/metrics/constants.go:29]: I think the convention would be `ConfigDBErrorsHelp`

**Review by calebdoxsey — APPROVED** [review]

File: .vscode/launch.json (hunk 1 above)
Comment (wasaga) [reply .vscode/launch.json:16]: otherwise it picks up random envoy from $PATH
what is your suggestion?

File: pkg/grpc/databroker/syncer.go (hunk 2 above)
Comment (wasaga) [reply pkg/grpc/databroker/syncer.go:158]: I used the same log level as i.e. 
```log
{"level":"info","service":"databroker","peer":"127.0.0.1:53995","type":"type.googleapis.com/pomerium.config.Config","id":"dashboard","time":"2021-04-06T15:13:46-04:00","message":"get"}
```

Comment (calebdoxsey) [reply pkg/grpc/databroker/syncer.go:158]: Well its going to be a lot of log lines. In our case ~40k?

File: .vscode/launch.json (hunk 1 above)
Comment (travisgroth) [reply .vscode/launch.json:16]: This has the OS and version hardcoded; I'm not sure if it will even work well on M1 chips since getenvoy doesn't support arm.

Would prefer something this opinionated isn't in a shared config.  Maybe it just shouldn't be in the repo, or we should make a more consistent location for the envoy binary.  We should tackle that separately.

**Review by travisgroth — CHANGES_REQUESTED** [review]

File: .vscode/launch.json [hunk 4]
```diff
@@ -10,7 +10,7 @@
       "args": [
         "-config",
         "${workspaceRoot}/.config.yaml"
-      ]
+      ],
```
Comment (desimone) [inline .vscode/launch.json:13]: nit: unrelated change.

File: pkg/grpc/databroker/syncer.go (hunk 2 above)
Comment (desimone) [reply pkg/grpc/databroker/syncer.go:158]: I agree this is a better fit for debug. 

And in this case, we may want to even hide this under the verbose flag. 

And we can follow up with another PR to see that databroker set / gets are also debug.

File: pkg/metrics/constants.go [hunk 5]
```diff
@@ -19,6 +19,14 @@ const (
 	PolicyCountTotal = "policy_count_total"
 	// ConfigChecksumDecimal should only be used to compare config on a single node, it will be different in multi-node environment
 	ConfigChecksumDecimal = "config_checksum_decimal"
+	// ConfigDBVersion sets currently loaded databroker config version config_db_version{service="service",id="config_id"}
+	ConfigDBVersion = "config_db_version"
+	// ConfigDBVersionHelp
```
Comment (desimone) [inline pkg/metrics/constants.go:24]: missing doctring.

File: pkg/metrics/constants.go [hunk 6]
```diff
@@ -19,6 +19,14 @@ const (
...
+	ConfigDBVersionHelp = "databroker current config record version"
+	// ConfigDBErrors sets number of errors while parsing current config that were tolerated
+	ConfigDBErrors = "config_db_errors"
+	// ConfigDBErrorsHelp
```
Comment (desimone) [inline pkg/metrics/constants.go:28]: missing docstring

File: internal/databroker/config_source.go [hunk 7]
```diff
@@ -84,40 +89,50 @@ func (src *ConfigSource) rebuild(firstTime bool) {
...
+	for id, cfgpb := range src.dbConfigs {
 		cfg.Options.ApplySettings(cfgpb.Settings)
+		var errCount uint64
 
 		err := cfg.Options.Validate()
 		if err != nil {
-			log.Warn().Err(err).Msg("databroker: invalid config detected, ignoring")
-			return
+			metrics.SetDBConfigRejected(cfg.Options.Services, id, cfgpb.version, err)
+			continue
```
Comment (desimone) [inline internal/databroker/config_source.go:99]: do we not want to fail fast in this case?

Comment (desimone) [reply internal/databroker/config_source.go:99]: Could this cause unexpected issues since we typically fail hard after failed validation?

**Review by desimone — CHANGES_REQUESTED** [review]

File: pkg/metrics/constants.go (hunk 6 above)
Comment (wasaga) [reply pkg/metrics/constants.go:28]: a help text for a variable with help text ? :-)

File: internal/databroker/config_source.go (hunk 7 above)
Comment (wasaga) [reply internal/databroker/config_source.go:99]: it loops over available databroker configs. idk what other configs were originally planned to exist besides one supplied by the console, but as long as we're looping over and merging them, it should be more natural to just proceed to next one.

File: pkg/metrics/constants.go (hunk 6 above)
Comment (desimone) [reply pkg/metrics/constants.go:28]: lol, that's fair. Maybe delete the docstring title then?

File: pkg/metrics/constants.go (hunk 5 above)
Comment (desimone) [reply pkg/metrics/constants.go:24]: https://github.com/pomerium/pomerium/pull/2065#discussion_r608330400

File: pkg/metrics/constants.go (hunk 6 above)
Comment (wasaga) [reply pkg/metrics/constants.go:28]: then linter would be unhappy

**Review by travisgroth — APPROVED** [review]

File: internal/databroker/config_source.go (hunk 7 above)
Comment (wasaga) [reply internal/databroker/config_source.go:99]: by returning we are ignoring other configs that may be valid. your logic is valid in case we have just a single config, however by design we have multiple here that we merge. 

@calebdoxsey please advise

Comment (wasaga) [reply internal/databroker/config_source.go:99]: also looping over map is random order that makes it even more non-deterministic. 
maybe we should eliminate the multiple databroker configs altogether here

Comment (calebdoxsey) [reply internal/databroker/config_source.go:99]: I don't know what to do if config is invalid.

Comment (travisgroth) [reply internal/databroker/config_source.go:99]: Was discussing with @wasaga - 

My understanding of the current code is that we're build a new config from the databroker by starting with a copy of current state, then mutating this copy.  It looks like we abort if the overall config doesn't validate, though we continue if individual routes have issues.  Is that right?

What I think I'd _want_ is to abort making changes if the new configuration is invalid for any reason.  If we proceed through a partial validation error, a previously functioning route or option could get dropped and cause an outage or otherwise undesirable state.  Preventing me from making new changes would retain whatever the current (likely functional) state is.  This does block new changes from happening, but I think that's the least surprising result and matches `HUP` behavior in a lot of other software.

File: pkg/metrics/constants.go (hunk 6 above)
Comment (desimone) [inline pkg/metrics/constants.go:28]:
```suggestion
	// ConfigDBErrorsHelp is the help text for ConfigDBErrors.
```

File: pkg/metrics/constants.go (hunk 5 above)
Comment (desimone) [inline pkg/metrics/constants.go:24]:
```suggestion
	// ConfigDBVersionHelp is the help text for ConfigDBVersion.
```

File: internal/databroker/config_source.go (hunk 7 above)
Comment (wasaga) [reply internal/databroker/config_source.go]: ok I'm reverting to original behaviour

File: internal/databroker/config_source.go [hunk 8]
```diff
@@ -84,40 +89,50 @@ func (src *ConfigSource) rebuild(firstTime bool) {
 	var additionalPolicies []config.Policy
 
 	// add all the config policies to the list
-	for _, cfgpb := range src.dbConfigs {
+	for id, cfgpb := range src.dbConfigs {
 		cfg.Options.ApplySettings(cfgpb.Settings)
+		var errCount uint64
```
Comment (desimone) [inline internal/databroker/config_source.go:94]: nit: `errCount` is now a ways away from its use. We might want to move this to just above the for loop

**Review by desimone — APPROVED** [review]
